File: yt/yt/flow/examples/python/url_downloader/url_download_function.py
"""UrlDownloadFunction: RowFunction processing URL download events."""

import logging
import time

from yt.yt.flow.library.python.companion.computation import RowFunction

log = logging.getLogger(__name__)


# [BEGIN url_download_function]
class UrlDownloadFunction(RowFunction):
    """Groups URLs by host, processes them with a timer-driven loop."""

    def on_message(self, message, output, ctx):
        host = message.payload["host"]
        url = message.payload["url"]

        state = ctx.state("host-state", message)
        data = state.get_or_default({"host": host, "pending_urls": []})
        data["pending_urls"].append(url)
        state.set(data)

        output.add_timer(int(time.time()) + 5)

    def on_timer(self, timer, output, ctx):
        state = ctx.state("host-state", timer)
        data = state.get_or_default(None)
        if not data or not data.get("pending_urls"):
            state.clear()
            return

        host = data["host"]
        for url in list(data["pending_urls"]):
            result = _process_url(url)
            builder = ctx.message_builder("processed_urls")
            builder.set("host", host)
            builder.set("url", url)
            builder.set("data", result)
            output.add_message(builder.finish())
            log.debug("Processed url (Host: %s, Url: %s)", host, url)

        state.clear()
# [END url_download_function]


def _process_url(url: str) -> str:
    digits = sum(1 for c in url if c.isdigit())
    return f"length: {len(url)}, digits: {digits}"

File: yt/yt/flow/examples/go/url_downloader/url_download_function.go
package main

import (
	"context"
	"fmt"
	"time"
	"unicode"

	"go.ytsaurus.tech/yt/go/flow"
)

const hostStateName = "host-state"

const outputStreamID = "processed_urls"

const flushDelay = 5 * time.Second

// [BEGIN host_state]

type hostState struct {
	Host        string   `yson:"host"`
	PendingURLs []string `yson:"pending_urls"`
}

// [END host_state]

// [BEGIN url_download_function]

type urlDownloadFunction struct{}

var (
	_ flow.RowFunction      = (*urlDownloadFunction)(nil)
	_ flow.RowTimerFunction = (*urlDownloadFunction)(nil)
)

func (*urlDownloadFunction) OnMessage(
	ctx context.Context,
	rt flow.Runtime,
	msg flow.ExtendedMessage,
	out flow.OutputCollector,
) error {
	var input urlMessage
	if err := msg.ConvertTo(&input); err != nil {
		return err
	}

	state, err := flow.OpenYSONState[hostState](rt, hostStateName, msg)
	if err != nil {
		return err
	}

	fresh := state.Empty()
	batch := state.Value()
	if fresh {
		batch.Host = input.Host
	}
	batch.PendingURLs = append(batch.PendingURLs, input.URL)

	out.AddTimer(flow.TimerRequest{TriggerTimestamp: uint64(time.Now().Add(flushDelay).Unix())})
	return nil
}

func (*urlDownloadFunction) OnTimer(
	ctx context.Context,
	rt flow.Runtime,
	timer flow.Timer,
	out flow.OutputCollector,
) error {
	state, err := flow.OpenYSONState[hostState](rt, hostStateName, timer)
	if err != nil {
		return err
	}

	if state.Empty() {
		return nil
	}
	batch := state.Value()
	if len(batch.PendingURLs) == 0 {
		state.Clear()
		return nil
	}

	for _, url := range batch.PendingURLs {
		processed := flow.NewYSONMessage[processedURLMessage](outputStreamID)
		processed.Host = batch.Host
		processed.URL = url
		processed.Data = processURL(url)
		encoded, err := flow.ConvertFrom(rt, processed)
		if err != nil {
			return err
		}
		out.AddMessage(encoded)
	}

	state.Clear()
	return nil
}

// [END url_download_function]

func processURL(url string) string {
	digits := 0
	for _, r := range url {
		if unicode.IsDigit(r) {
			digits++
		}
	}
	return fmt.Sprintf("length: %d, digits: %d", len(url), digits)
}
